test/queue.js

```diff
@@ -188,4 +209,85 @@ describe("Queue methods", function() {
             checkDone();
         });
     });
+
+    it("should update items in the queue", function(done) {
+        crawler.queue.update("http://127.0.0.1:3000/stage2", {
+            status: "queued",
+            fetched: false
+        }, function(error, queueItem) {
+            queueItem.should.include({
+                url: "http://127.0.0.1:3000/stage2",
+                status: "queued",
+                fetched: false
+            });
+
+            done(error);
+        });
+    });
+
+    /**
+     * This test works by monkey patching the queue `add` and `update` methods
+     * and keeping a local copy of the queue, which contains cloned queueItems.
+     * Each time the `update` method is called, we deeply compare the copy in
+     * the queue with our local one. Same thing when the crawler completes.
+     */
+    it("should only update queue items asynchronously", function(done) {
+        var crawler = new Crawler("http://127.0.0.1:3000"),
+            originalQueueAdd = crawler.queue.add,
+            originalQueueUpdate = crawler.queue.update;
+
+        var queueItems = [];
+
+        crawler.interval = 5;
+        crawler.maxDepth = 2;
+
+        function findByUrl(array, url) {
+            return find(array, function(queueItem) {
+                return queueItem.url === url;
+            });
+        }
+
+        crawler.queue.add = function (queueItem) {
+            var storedQueueItem = deepAssign({}, queueItem);
```

Comment:
Sure, we could wrap it in some setImmediate calls, but is there any particular scenario where we run the risk of a race condition here.?